src/osd/OSD.cc

```diff
@@ -1718,10 +1726,27 @@ void OSDService::send_ready_to_merge()
 
 void OSDService::_send_ready_to_merge()
 {
+  for (auto src : not_ready_to_merge_source) {
+    if (sent_ready_to_merge_source.count(src) == 0) {
+      monc->send_mon_message(new MOSDPGReadyToMerge(
+			       src,
+			       0,
+			       false,
+			       osdmap->get_epoch()));
+    }
+  }
   for (auto src : ready_to_merge_source) {
-    if (ready_to_merge_target.count(src.get_parent()) &&
+    if (not_ready_to_merge_source.count(src)) {
+      continue;
+    }
+    auto p = ready_to_merge_target.find(src.get_parent());
```

Comment:
This looks ok for now, but I wonder if we can do a more complete verification here to confirm that a merge parent and all its children are ready and then sending out the Ready message.  Is there any plan we are going to support batch merge later?

Reply:
No plans for batch merge.

The only additional check i can think of here is on the monitor: verify that the pool hasn't changed when processing the ReadyToMerge message and, if it has, revert the merge instead of completing it.  The trick is doing that while avoiding checking all intervenining OSDMaps.  I think I can do it with the change epoch in pg_pool_t and an imprecise check for the OSD mappings, since any CRUSH etc changes generally shouldn't matter (pgp_num is small so the source and target PGs should always map to the same place, barring a pg_temp or upmap entry).

Reply:
I kind of want to run it without that check for a while though to get some confidence that the OSDs can handle it first...